readme.md
# BUSES SPEECH NOTIFIER

![](bus.png) SPEECH NOTIFICATION OF THE NEXT BUS SCHEDULED (powershell powered)

## INTRO 

Every morning my wife ask to me for the next buses to start its optimized morning run....Huuum this is a repetive task that can be automated with some lines of Powershell. We live in Paris (France) area and the bus service api requested is [Transdev-idf](https://www.transdev-idf.com/info-trafic) but its easy to convert it to another city or project, for example to notify about anything...a new commit...

![](bus_notifier.png)

## APPLICATION FLOWS

```mermaid
flowchart LR;
    B[APPLICATION]-->|schedule for my bus line please| A[Transdev bus API];
    A[Transdev bus API]-.->|json schedule| B[APPLICATION];
    B[APPLICATION]-->|json schedule| J[JQ json parser];
    J[JQ json parser]-.->|json schedule of interest| B[APPLICATION]
    B[APPLICATION]-->|next bus in...minutes| V[Vocal module];
    style B fill:#FFD700,stroke:#333,stroke-width:4px
```

## TECH STACK

- Powershell Core 7.2.1 on Windows
    >https://github.com/PowerShell/PowerShell/releases

- [jq - A command-line JSON processor](https://stedolan.github.io/jq/)
    >chocolatey install jq
    
    >curl 'https://api.github.com/repos/stedolan/jq/commits?per_page=5' | jq '.[] | {message: .commit.message, name: .commit.committer.name}'
    
    >A jq playground to test jq syntax: [jqplay](https://jqplay.org/)

## USAGE

>PS>./r1_bus_notifier.ps1

>CMD>pwsh ./r1_bus_notifier.ps1

#### With Task-Scheduler

```ps
# Register the new PowerShell scheduled task

# The name of your scheduled task.
$taskName = "Bus_R1_Notifier"
# Describe the scheduled task.
$description = "Notify #R1 bus schedule"
$taskTrigger = New-ScheduledTaskTrigger -Weekly -DaysOfWeek Monday, Tuesday, Wednesday, Thursday, Friday -At "07:20AM"
$taskAction = New-ScheduledTaskAction `
    -Execute 'pwsh.exe' `
    -Argument "-File K:\tools\bus_notifier\r1_bus_notifier.ps1" -WorkingDirectory "K:\tools\bus_notifier"

# Register the scheduled task
Register-ScheduledTask `
    -TaskName $taskName `
    -Action $taskAction `
    -Trigger $taskTrigger `
    -Description $description
```

To test the scheduled task, you can either wait for the next run schedule or use the command below to manually trigger.
>Start-ScheduledTask -TaskName Bus_R1_Notifier

>Unregister-ScheduledTask -TaskName 'Bus_R1_Notifier' -Confirm:$false


### TECHNICAL DETAILS

You can find more details about the api used [here](technical_details.md)

## CONTRIBUTING

Wow, thanks to contribute to this project!   
New ideas, samples, forks...are welcome.  
So please just fork it, commit, push and send me a pull request to main.  

## CHANGELOG

See [CHANGELOG](CHANGELOG.md) for more details.

## LICENSE

This project is licensed under the [MIT License](LICENSE).

## ROADMAP

- ✔️ Make window nice
- ❌ Improve the speech
- ❌ Warn when service is unavailable.
- ❌ Detect shift in schedule from day to day and warn the user.
- ❌ Add connections to others transportation services (RER A for Paris area) to say 'connection with train x at 00h00 possible'.

### NOTES

Bus image comes from https://emojipedia.org/bus/

As jq for parsing Json, you can also try [htmlq](https://github.com/mgdm/htmlq), a jq but for HTML. Uses CSS selectors to extract bits of content from HTML files.

### DONATE

You can also contribute to the continued success of the project via donation. Please click this link.


CHANGELOG.md
# CHANGELOG

<!--
Read this section if it's your first time writing changelog, if not read anyway.

Guidelines:
- Don't dump commit log diffs as changelogs. Bad idea, it is.
- Changelogs are for humans, not machines.
- There should be an entry for every single version.
- The same types of changes should be grouped.
- the latest version comes first.

Tags:
- Added: for new features.
- Changed: for changes in existing functionality.
- Deprecated: for soon-to-be removed features.
- Removed: for now removed features.
- Fixed: for any bug fixes.
- Security: in case of vulnerabilities.

Good to have: commit or PR links.

-->

## v1.0.1

### Changed

- Start vocal notification immediately at startup (before we have to wait for the interval)

## v1.0.0

### Changed

- Use pure XAML as window definition
- Add images  
- Review README
- Add CHANGELOG

## v0.0.1 

- POC, testing jq library


vocal_message.ps1
# REQUIRED LIBRARIES
Add-Type -AssemblyName System.speech;       # SPEECH SYNTHETIS
. ./apiservice.ps1      # if executed from task-scheduler... define full path here for the .ps1 file OR specifying the running directory for scheduled task (-WorkingDirectory argument)

function getSchedules() {

    # GET DATA AS JSON
    $apiService = [ApiService]::new($apiParams)
    $apiJsonResponse = ConvertTo-Json $apiService.Invoke() 
    
    # FILTER DATA
    
    $dataFilter = "[.[]| select(.station==$stationOrigin and .destination==$stationDestination" + ')] | group_by(.line, .updated) | map({ "line": .[0].line, "updated": .[0].updated, "delays_minutes": map(.delay/60) | unique })'
    $filteredResponse = $apiJsonResponse | jq $dataFilter
    
    return $filteredResponse
    <# [ {
        "line": "R1",
        "updated": "2022-02-22T17:04:43",
        "delays_minutes": [ 1.2833333333333334, 7.283333333333333 ] } ]
     #>
}

function vocalNotification($schedules) {
    
    $nextBuses = $schedules | ConvertFrom-Json

    $speaker = New-Object System.Speech.Synthesis.SpeechSynthesizer;     
    foreach ($bus in $nextBuses)
    {
        write-host "  Line $($bus.line)"
        $speaker.Speak("Next bus to Saint Germain-en-Laye, line $($bus.line.split('')) in");
        $bus.delays_minutes | foreach {
            $nextDeparture = [System.Math]::Floor($_)
            $sentence = ($nextDeparture -lt 1) ? "less than one minute" : "$($nextDeparture) minutes";
            write-host "   🚌 $sentence"
            $speaker.Speak($sentence);
        }
    }
    $speaker.Dispose()
}

# API CALL PARAMETERS
$apiParams = @{   
    contentType = 'application/json';
    method = 'POST';
    body = 'token=undefined';
    uri = 'https://www.transdev-idf.com/ajax/station/540445338/nextbus';
    headers = @{ 'Host' = 'www.transdev-idf.com'
        'Accept' = 'application/json, text/javascript, */*; q=0.01'
        'X-Requested-With' = 'XMLHttpRequest'
        'Origin' = 'https://www.transdev-idf.com'
        'Accept-Encoding' = 'gzip, deflate, br'};
}

# Data Filter Parameters
$stationOrigin = "540445338"        # Lycée Leonard de Vinci
$stationDestination = "50012439"    # Saint Germain en Laye RER

$schedules = getSchedules
vocalNotification($schedules)


r1_bus_notifier.ps1
#
# BUS NOTIFIER V1.0.1
# https://github.com/guinetn/bus_notifier/edit/main/readme.md
#

# REQUIRED LIBRARIES
Add-Type -AssemblyName PresentationFramework, PresentationCore, WindowsBase

[xml]$xamlMainWindow = @"
<Window 
xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"     
SizeToContent="WidthAndHeight" Title="BUS NOTIFIER 🚌">
    <Border Background="#FFD700" BorderBrush="Gray" BorderThickness="2">
        <Grid>
            <Image Source="$PSScriptRoot\bus.png" Stretch="None" Margin="-5,30,0,-20" HorizontalAlignment="Left"/>
            <StackPanel Margin="10">
            <TextBlock FontSize="14" ToolTip="Click to open timetable source">
                    <Hyperlink Name="linkToTimeTable" 
                            NavigateUri="https://www.transdev-idf.com/horaires-ligne-R1/lycee-de-vinci-vers-gare-de-st-germain-en-laye/012-RESALY-540445338-50012439">
                            BUS R1 ▶▶ SAINT-GERMAIN RER</Hyperlink>
                </TextBlock>
                <Label FontWeight="Bold" Content="{Binding Path=Value, ElementName=sliderNotificationFrequency}" ContentStringFormat="Vocal Information Frequency: {0:###} seconds" />
                <Slider Name="sliderNotificationFrequency" Value="60" Maximum="600" Minimum="30" TickFrequency="30" IsSnapToTickEnabled="True" TickPlacement="BottomRight"></Slider>
                <Button Name="buttonCloseApp" Content="Close" Background="#000000" Foreground="#FFF" Margin="10" HorizontalAlignment="Right" Width="100px" Height="40px"></Button>
            </StackPanel> 
        </Grid>
    </Border>
</Window>
"@

$timerNotification = [System.Windows.Threading.DispatcherTimer]::new()

$MainFunction = {

    $window = CreateWindow
    GenerateVariables
    RegisterEvents
    StartNotifications
    NotifyUser # notify user immediately
    
    $window.ShowDialog() | Out-Null
    UnRegisterEvents
 }

 function CreateWindow {
    $xmlReader=(New-Object System.Xml.XmlNodeReader $xamlMainWindow)
    return [Windows.Markup.XamlReader]::Load( $xmlReader )
}

function GenerateVariables() {
    # Create variables from xaml nodes having a 'NAME' attribute. Make them script scoped
    $xamlMainWindow.SelectNodes("//*[@Name]") | ForEach-Object {Set-Variable -Scope Script -Name ($_.Name) -Value $window.FindName($_.Name)}
}

function UnRegisterEvents() {
    $timerNotification.Stop()
}

function RegisterEvents() {
    
    $linkToTimeTable.add_Click({
        Start-Process $linkToTimeTable.NavigateUri;
    })

    $sliderNotificationFrequency.add_valuechanged({ 
        $actionFrequency = [int]$sliderNotificationFrequency.value
        $timerNotification.Interval = [Timespan]::FromSeconds($actionFrequency)
    }) 

    $buttonCloseApp.add_Click({ 
        UnRegisterEvents
        $window.Close() 
    }) 
}

function StartNotifications {
  
    $timerNotification.Interval =  [Timespan]::FromSeconds([int]$sliderNotificationFrequency.value)
    $timerNotification.Add_Tick( { 
        NotifyUser
    } )
    $timerNotification.Start()
}

function NotifyUser {
    write-host  "$(Get-Date -Format "yyyy-MM-dd HH:mm:ss") Vocal notification" 
    $actionScript = Join-Path $PSScriptRoot -ChildPath "vocal_message.ps1"
    Start-ThreadJob -FilePath $actionScript 
}

& $MainFunction


aural_message.ps1
# REQUIRED LIBRARIES
Add-Type -AssemblyName System.speech;       # SPEECH SYNTHETIS
. ./apiservice.ps1

function getSchedules() {

    # GET DATA AS JSON
    $apiService = [ApiService]::new($apiParams)
    $apiJsonResponse = $apiService.Invoke() |  Select-Object -Property * | ConvertTo-Json 
    
    # FILTER DATA
    
    $dataFilter = "[.[]| select(.station==$stationOrigin and .destination==$stationDestination" + ')] | group_by(.line, .updated) | map({ "line": .[0].line, "updated": .[0].updated, "delays_minutes": map(.delay/60) | unique })'
    $filteredResponse = $apiJsonResponse | jq $dataFilter
    
    return $filteredResponse
    <# [ {
        "line": "R1",
        "updated": "2022-02-22T17:04:43",
        "delays_minutes": [ 1.2833333333333334, 7.283333333333333 ] } ]
     #>
}

function auralMessage($schedules) {
    
    $nextBuses = $schedules | ConvertFrom-Json

    $speaker = New-Object System.Speech.Synthesis.SpeechSynthesizer;     
    foreach ($bus in $nextBuses)
    {
        write-host "  Line $($bus.line)"
        $speaker.Speak("Next bus to Saint Germain-en-Laye, line $($bus.line.split('')) in");
        $bus.delays_minutes | foreach {
            $nextDeparture = [System.Math]::Floor($_)
            $sentence = ($nextDeparture -lt 1) ? "less than one minute" : "$($nextDeparture) minutes";
            write-host "   🚌 $sentence"
            $speaker.Speak($sentence);
        }
    }
}

# API CALL PARAMETERS
$apiParams = @{   
    contentType = 'application/json';
    method = 'POST';
    body = 'token=undefined';
    uri = 'https://www.transdev-idf.com/ajax/station/540445338/nextbus';
    headers = @{ 'Host' = 'www.transdev-idf.com'
        'Accept' = 'application/json, text/javascript, */*; q=0.01'
        'X-Requested-With' = 'XMLHttpRequest'
        'Origin' = 'https://www.transdev-idf.com'
        'Accept-Encoding' = 'gzip, deflate, br'};
}

# Data Filter Parameters
$stationOrigin = "540445338"        # Lycée Leonard de Vinci
$stationDestination = "50012439"    # Saint Germain en Laye RER

$schedules = getSchedules
auralMessage($schedules)



<# BUS SERVICE API: DETAILS

Api      https://www.transdev-idf.com/horaires-ligne-R1/lycee-de-vinci-vers-gare-de-st-germain-en-laye/012-RESALY-540445338-50012439
Extract  Next buses departures

API REQUEST
    POST   https://www.transdev-idf.com/ajax/station/540445338/nextbus
    BODY   token:undefined    !! MANDATORY, AS IT IS !!
    HEADERS
        Host=www.transdev-idf.com;
        Accept=application/json, text/javascript, */*; q=0.01
        X-Requested-With=XMLHttpRequest
        Origin=https://www.transdev-idf.com
        Accept-Encoding=gzip, deflate, br

    All HEADERS before headers cleanup: some are not essential
        Host: www.transdev-idf.com
        Connection: keep-alive
        Content-Length: 15
        sec-ch-ua: " Not A;Brand";v="99", "Chromium";v="98", "Google Chrome";v="98"
        Accept: application/json, text/javascript, */*; q=0.01
        Content-Type: application/x-www-form-urlencoded; charset=UTF-8
        X-Requested-With: XMLHttpRequest
        sec-ch-ua-mobile: ?0
        User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/98.0.4758.102 Safari/537.36
        sec-ch-ua-platform: "Windows"
        Origin: https://www.transdev-idf.com
        Sec-Fetch-Site: same-origin
        Sec-Fetch-Mode: cors
        Sec-Fetch-Dest: empty
        Referer: https://www.transdev-idf.com/horaires-ligne-R1/lycee-de-vinci-vers-gare-de-st-germain-en-laye/012-RESALY-540445338-50012439
        Accept-Encoding: gzip, deflate, br
        Accept-Language: fr,en;q=0.9

        
TEST
    $params = @{   
        contentType = 'application/json';
        method = 'POST';
        body = 'token=undefined';
        uri = 'https://www.transdev-idf.com/ajax/station/540445338/nextbus';
        headers = @{ 'Host' = 'www.transdev-idf.com'
            'Accept' = 'application/json, text/javascript, */*; q=0.01'
            'X-Requested-With' = 'XMLHttpRequest'
            'Origin' = 'https://www.transdev-idf.com'
            'Accept-Encoding' = 'gzip, deflate, br'};
    }
    $ret = Invoke-RestMethod @params 
    write-output $ret


RESPONSE
    
    Next bus in  ...  8 min
                 ... 23 min  
        Updated at 12h51 

    [
    {
        "charge": -1,
        "station": 540445338,
        "line": "R1",
        "delay": 525,                           525 / 60 = 8.75 min     ← 1st next bus at... 
        "direction": "backward",
        "wheelchair": true,
        "source": "INEO_RT",
        "updated": "2022-02-22T12:51:15",       ← updated at...
        "destination": 50012439
    },
    {
        "charge": -1,
        "station": 540445338,
        "line": "R1",
        "delay": 1425,                      1425 / 60 = 23 min      ← next bus at...
        "direction": "backward",
        "wheelchair": true,
        "source": "INEO_RT",
        "updated": "2022-02-22T12:51:15",
        "destination": 50012439
    },
    {
        "charge": -1,
        "station": 540445338,
        "line": "R6",                       ← not interested in the R6 line
        "delay": 1725,
        "direction": "backward",
        "wheelchair": true,
        "source": "INEO_RT",
        "updated": "2022-02-22T12:51:15",
        "destination": 50012444
    },
    {
        "charge": -1,
        "station": 540445338,
        "line": "R6",                       ← not interested in the R6 line
        "delay": 5325,
        "direction": "backward",
        "wheelchair": false,
        "source": "INEO_RT",
        "updated": "2022-02-22T12:51:15",
        "destination": 50012444
    }
    ]



## JQ: JSON processor

- Command-line JSON processor: https://stedolan.github.io/jq/
- A playground for jq, written in Go: https://jqplay.org/


INPUT

    [
    {
        "source": "INEO_RT",
        "updated": "2022-02-22T13:00:31",
        "direction": "backward",
        "charge": -1,
        "line": "R1",
        "destination": 50012439,
        "delay": 869,
        "wheelchair": true,
        "station": 540445338
    },
    {
        "source": "INEO_RT",
        "updated": "2022-02-22T13:00:31",
        "direction": "backward",
        "charge": -1,
        "line": "R1",
        "destination": 50012439,
        "delay": 1769,
        "wheelchair": true,
        "station": 540445338
    },
    {
        "source": "INEO_RT",
        "updated": "2022-02-22T13:00:31",
        "direction": "backward",
        "charge": -1,
        "line": "R6",
        "destination": 50012444,
        "delay": 1169,
        "wheelchair": true,
        "station": 540445338
    },
    {
        "source": "INEO_RT",
        "updated": "2022-02-22T13:00:31",
        "direction": "backward",
        "charge": -1,
        "line": "R6",
        "destination": 50012444,
        "delay": 4769,
        "wheelchair": false,
        "station": 540445338
    }
    ]

FILTERS
    keep line R1 from bus station 540445338 to 50012439 

JQ FILTER 1

    .[] | select((.line== "R1") and .destination==50012439 and .station==540445338) | {line,  delay, delay_min:(.delay/60), updated:.updated}

    RESULT

        {
        "line": "R1",
        "delay": 869,
        "delay_min": 14.483333333333333,
        "updated": "2022-02-22T13:00:31"
        }
        {
        "line": "R1",
        "delay": 1769,
        "delay_min": 29.483333333333334,
        "updated": "2022-02-22T13:00:31"
        }


JQ FILTER 2

    [ .[] | select((.line== "R1") and .destination==50012439 and .station==540445338) |  {line,  delay, delay_min:(.delay/60), updated:.updated} ] | group_by(.line) | map({ "line": .[0].line, "delays": map(.delay) | unique }) 

    RESULT
    
        [
        {
            "line": "R1",
            "delays": [
            0.45,
            15.45
            ]
        }
        ]

JQ FILTER 3

    [ .[] | select(.destination==50012439 and .station==540445338)] |  group_by(.line, .updated) | map({ "line": .[0].line, "updated": .[0].updated, "delays_minutes": map(.delay/60) | unique }) 
#>